lib/ansible/modules/cloud/openstack/os_server_metadata.py

```diff
@@ -0,0 +1,197 @@
+          the instance or a string containing a list of key-value pairs.
+          Eg:  meta: "key1=value1,key2=value2"'
+     required: true
+   state:
+     description:
+       - Should the resource be present or absent.
+     choices: [present, absent]
+     default: present
+   availability_zone:
+     description:
+       - Availability zone in which to create the snapshot.
+     required: false
+requirements:
+    - "python >= 2.7"
+    - "shade"
+'''
+
+EXAMPLES = '''
+# Creates or updates hostname=test1 as metadata of the server instance vm1
+- name: add metadata to compute instance
+  hosts: localhost
+  tasks:
+  - name: add metadata to instance
+    os_server_metadata:
+        state: present
+        auth:
+            auth_url: https://openstack-api.example.com:35357/v2.0/
+            username: admin
+            password: admin
+            project_name: admin
+        name: vm1
+        meta: "hostname=test1,group=group1"
+
+# Removes the keys under meta from the instance named vm1
+- name: delete metadata from compute instance
+  hosts: localhost
+  tasks:
+  - name: delete metadata from instance
+    os_server_metadata:
+        state: absent
+        auth:
+            auth_url: https://openstack-api.example.com:35357/v2.0/
+            username: admin
+            password: admin
+            project_name: admin
+        name: vm1
+        meta: "hostname=,group="
+'''
+
+RETURN = '''
+server_id:
+    description: The compute instance id where the change was made
+    returned: success
+    type: string
+    sample: "324c4e91-3e03-4f62-9a4d-06119a8a8d16"
+metadata:
+    description: The metadata of compute instance after the change
+    returned: success
+    type: dict
+    sample: {'key1': 'value1', 'key2': 'value2'}
+'''
+
+try:
+    import shade
+    HAS_SHADE = True
+except ImportError:
+    HAS_SHADE = False
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.openstack import (openstack_full_argument_spec,
+                                            openstack_module_kwargs)
+
+
+def _needs_update(server_metadata=None, metadata=None):
+    if server_metadata is None:
+        server_metadata = {}
+    if metadata is None:
+        metadata = {}
+    return len(set(metadata.items()) - set(server_metadata.items())) != 0
+
+
+def _get_keys_to_delete(server_metadata_keys=None, metadata_keys=None):
+    if server_metadata_keys is None:
+        server_metadata_keys = []
+    if metadata_keys is None:
+        metadata_keys = []
+    return set(server_metadata_keys) & set(metadata_keys)
+
+
+def _csv_to_dict(meta_param):
+    metas = {}
+    for kv_str in meta_param.split(","):
+        k, v = kv_str.split("=")
+        metas[k.strip()] = v
+    return metas
+
+
+def main():
+    argument_spec = openstack_full_argument_spec(
+        server=dict(required=True, aliases=['name']),
+        meta=dict(required=True, type='str'),
```

Comment:
Would it be possible to have this field accept a dict instead of `k=v` string?

```yaml
- os_server_metadata:
    server: vm1
    state: absent
    meta:
      hostname: ''
      group: ''
```

Reply:
Had that before and changed it for no reason :)
Fixed.